src/NimBLEClient.cpp

```diff
@@ -989,15 +989,16 @@ int NimBLEClient::handleGapEvent(struct ble_gap_event* event, void* arg) {
                     break;
                 }
 
-                if (pClient->m_config.deleteOnConnectFail) { // async connect
+                // async connect
+                pClient->m_pClientCallbacks->onConnectFail(pClient, rc);
+
+                if (pClient->m_config.deleteOnConnectFail) {
                     delete pClient;
-                    return 0;
                 }
+                return 0;
             }
 
-            if (pClient->m_config.asyncConnect) {
-                pClient->m_pClientCallbacks->onConnect(pClient);
-            } else if (!pClient->m_config.exchangeMTU) {
```

Comment:
This call to `onConnect` should be left as is for async connects.

Reply:
Oops, I actually meant to move that instead of delete, thanks!